c++/src/capnp/compat/json.c++

```diff
@@ -372,6 +379,313 @@ Orphan<DynamicValue> JsonCodec::decode(
+            advance(4);
+            break;
+          default: KJ_FAIL_REQUIRE("Invalid escape in JSON string."); break;
+        }
+      }
+
+    } while(nextChar() != '"');
+
+    consume('"');
+    decoded.add('\0');
+
+    // TODO(perf): This copy can be eliminated, but I can't find the kj::wayToDoIt();
+    return kj::String(decoded.releaseAsArray());
+  }
+
+  kj::String consumeNumber() {
+    // TODO(perf): <regex> could be eliminated.
+    static std::regex jsonNumber(
+        R"(-?(0|[1-9][0-9]*)(\.[0-9]+)?((e|E)([+-])?[0-9]+)?)");
+    // jsonNumber matches numbers as seen in railroad diagram at http://json.org/.
+
+    std::cmatch match;
+    bool foundNumber = std::regex_search(
+        remaining_.begin(),
+        remaining_.end(),
+        match, jsonNumber,
+        std::regex_constants::match_continuous);
+
+    KJ_REQUIRE(foundNumber, "Expected number in JSON input.");
+
+    kj::Vector<char> number;
+    number.addAll(match[0].first, match[0].second);
+    number.add('\0');
+
+    advanceTo(match.suffix().first);
+
+    return kj::String(number.releaseAsArray());
+  }
+
+  // TODO(someday): This "interface" is ugly, and won't work if/when surrogates are handled.
+  void unescapeAndAppend(kj::ArrayPtr<const char> hex, kj::Vector<char>& target) {
+    KJ_REQUIRE(hex.size() == 4);
+    int codePoint = 0;
+
+    for (int i = 0; i < 4; ++i) {
+      char c = hex[i];
+      codePoint <<= 4;
+
+      if ('0' <= c && c <= '9') {
+        codePoint |= c - '0';
+      } else if ('a' <= c && c <= 'f') {
+        codePoint |= c - 'a';
+      } else if ('A' <= c && c <= 'F') {
+        codePoint |= c - 'A';
+      } else {
+        KJ_FAIL_REQUIRE("Invalid hex digit in unicode escape.", c);
+      }
+    }
+
+    // TODO(soon): Support at least basic multi-lingual plane, ie ignore surrogates.
+    KJ_REQUIRE(codePoint < 128, "non-ASCII unicode escapes are not supported (yet!)");
+    target.add(0x7f & static_cast<char>(codePoint));
+  }
+
+private:
+  static const kj::ArrayPtr<const char> NULL_;
+  static const kj::ArrayPtr<const char> FALSE;
+  static const kj::ArrayPtr<const char> TRUE;
+
+  const size_t maxNestingDepth_;
+  const kj::ArrayPtr<const char> input_;
+  kj::ArrayPtr<const char> remaining_;
+  size_t nestingDepth_;
+
+};  // class Parser
+
+// clang warns about these constructors running on program start. All they do is each set a pointer
+// and a size_t, so we politely ask clang to let it slide.
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wglobal-constructors"
+#endif
+
+// Array literal used instead of string literal to avoid null terminator.
+const kj::ArrayPtr<const char> Parser::NULL_ = kj::ArrayPtr<const char>({'n','u','l','l'});
+const kj::ArrayPtr<const char> Parser::FALSE = kj::ArrayPtr<const char>({'f','a','l','s','e'});
+const kj::ArrayPtr<const char> Parser::TRUE = kj::ArrayPtr<const char>({'t','r','u','e'});
+
```

Comment:
(That was a random driveby comment -- unfortunately I don't have time to review all your updates at the moment, but will try to get to it soon.)